idaes/apps/grid_integration/bidder.py

```diff
@@ -0,0 +1,774 @@
+
+        # ramp up limits
+        def ramp_up_fun(m,j,h,k):
+            '''
+            j,h stand for unit, hour,scenario respectively.
+            '''
+            if h==0:
+                return m.P_T[j,h,k] <= m.pre_P_T[j] \
+                + m.ramp_up[j]*m.pre_on_off[j]\
+                + m.ramp_start_up[j]*m.start_up[j,h,k]
+            else:
+                return m.P_T[j,h,k] <= m.P_T[j,h-1,k] \
+                + m.ramp_up[j]*m.on_off[j,h-1,k]\
+                + m.ramp_start_up[j]*m.start_up[j,h,k]
+        m.ramp_up_con = pyo.Constraint(m.UNITS,m.HOUR,m.SCENARIOS,rule = ramp_up_fun)
+
+        # ramp shut down limits
+        def ramp_shut_dw_fun(m,j,h,k):
+            '''
+            j,h stand for unit, hour,scenario respectively.
+            '''
+            if h==0:
+                return m.pre_P_T[j] <= m.Pmax[j]*m.on_off[j,h,k] + m.ramp_shut_dw[j] * m.shut_dw[j,h,k]
+            else:
+                return m.P_T[j,h-1,k] <= m.Pmax[j]*m.on_off[j,h,k] + m.ramp_shut_dw[j] * m.shut_dw[j,h,k]
+        m.ramp_shut_dw_con = pyo.Constraint(m.UNITS,m.HOUR,m.SCENARIOS,rule = ramp_shut_dw_fun)
+
+        # ramp down limits
+        def ramp_dw_fun(m,j,h,k):
+            '''
+            j,h stand for unit, hour,scenario respectively.
+            '''
+            if h == 0:
+                return m.pre_P_T[j] - m.P_T[j,h,k] <= m.ramp_dw[j] * m.on_off[j,h,k]\
+                + m.ramp_shut_dw[j] * m.shut_dw[j,h,k]
+            else:
+                return m.P_T[j,h-1,k] - m.P_T[j,h,k] <= m.ramp_dw[j] * m.on_off[j,h,k]\
+                + m.ramp_shut_dw[j] * m.shut_dw[j,h,k]
+        m.ramp_dw_con = pyo.Constraint(m.UNITS,m.HOUR,m.SCENARIOS,rule = ramp_dw_fun)
+
+        ## add min up and down time constraints
+        self._add_UT_DT_constraints(m)
+
+        ## add bidding constraints
+        self._add_bidding_constraints(m)
+
+        ## Expression
+
+        def prod_cost_fun(m,j,h,k):
+            return m.min_load_cost[j] * m.on_off[j,h,k] \
+            + sum(m.F[j,l]*m.power_segment[j,h,l,k] for l in m.SEGMENTS)
+        m.prod_cost_approx = pyo.Expression(m.UNITS,m.HOUR,m.SCENARIOS,rule = prod_cost_fun)
+
+        # start up costs
+        def start_cost_fun(m,j,h,k):
+            return m.start_up_cost[j]*m.start_up[j,h,k]
+        m.start_up_cost_expr = pyo.Expression(m.UNITS,m.HOUR,m.SCENARIOS,rule = start_cost_fun)
+
+        # total cost
+        def tot_cost_fun(m,j,h,k):
+            return m.prod_cost_approx[j,h,k] + m.start_up_cost_expr[j,h,k]
+        m.tot_cost = pyo.Expression(m.UNITS,m.HOUR,m.SCENARIOS,rule = tot_cost_fun)
+
+        ## Objective
+        def exp_revenue_fun(m):
+            return sum(m.P_T[j,h,k]*m.DAM_price[h,k]- m.tot_cost[j,h,k] for h in m.HOUR for j in m.UNITS for k in m.SCENARIOS)
+        m.exp_revenue = pyo.Objective(rule = exp_revenue_fun,sense = pyo.maximize)
+
+        return m
+
+    @staticmethod
+    def _update_UT_DT(m,implemented_shut_down, implemented_start_up):
```

Comment:
@jsiirola What are your thoughts on leaving this as a static method with `m` as an input or having `m` live in `self` and making this a normal class method?

Reply:
Same comment applies to multiple functions.